### ISSUE TITLE
different memory usage from :clone() versus :share()

### ISSUE THREAD

**Issue by willwhitney** [torch/nn#742]
According to the docs, `:clone`:

> Creates a deep copy of (i.e. not just a pointer to) the module, including the current state of its parameters (e.g. weight, biases etc., if any). If arguments are provided to the clone(...) function it also calls share(...) with those arguments on the cloned module after creating it, hence making a deep copy of this module with some shared parameters.

From that I'd expect that cloning a module with some arguments, and creating a second module and calling share with the same arguments to have the same memory usage. However, cloning with args doesn't seem to actually share the tensors.

Also, the peak memory usage of clone is absurd.

Here's share, which seems to have the right behavior:

```lua
a = nn.Linear(10000,10000)
-- memory usage: 766.0MB
b = nn.Linear(10000,10000)
-- memory usage: 1.49GB
b:share(a, 'weight', 'bias', 'gradWeight', 'gradBias')
-- memory usage: 766.2MB
```

and clone, which is really bad:

```lua
a = nn.Linear(10000,10000)
-- memory usage: 765.9MB
b = a:clone('weight', 'bias', 'gradWeight', 'gradBias')
-- memory usage: 2.98GB
collectgarbage()
-- memory usage: 1.49GB
```

Is this intended behavior? If so, we should probably update the docs with a huge warning label on clone.

**Comment by soumith** [comment]
I tried your examples, and in the second example, if you call collectgarbage(); collectgarbage(); it goes down back to ~770MB. I think the first time when collectgarbage() is called, some variables are not finalized yet.

**Comment by fmassa** [comment]
I second what @soumith said. The memory usage in the second example is initially higher because a [new temporary object `f` is created when calling `:clone`](https://github.com/torch/nn/blob/master/Module.lua#L105-L115), which is as big as the original object, but then it's cleaned when it becomes out of scope.

**Comment by fmassa** [comment]
(I have the impression though that `torch.MemoryFile` objects are not counting towards the heaptracking, making it necessary to call `collectgarbage()` if we want to free `t`)

**Comment by apaszke** [comment]
@fmassa Yes, it probably won't be freed. That could be probably fixed by adding something like `THFile_truncate()` and exposing it to lua, so it cold be called there.

**Comment by willwhitney** [comment]
@soumith On OSX a second call to collectgarbage() doesn't drop my memory usage. It stays at 1.49GB. But it looks like this does the right thing on Linux, which is what actually matters to me, so this isn't really a priority.
